### DISCUSSION TITLE
After implemented RWS, still get `prompt` state and need `user gesture` to get permission?

### DISCUSSION THREAD

**Discussion by lqw1111** [privacysandbox/privacy-sandbox-dev-support#175] [Phasing out third-party cookies] [answered]
Hi,
A question about Related websites sets and 3rd party cookies.

**Context**:
we have two websites, `web.com` and `iframe.com`, and `iframe.com` was embeded in the `web.com` as iframe, and `web.com` will `postMessage` to `iframe.com`
I am running chrome using this command 
```
/Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --enable-features=FirstPartySets,StorageAccessAPI,StorageAccessAPIForOriginExtension,PageInfoCookiesSubpage,PrivacySandboxFirstPartySetsUI --test-third-party-cookie-phaseout --use-first-party-set="{\"primary\": \"https://web.com\", \"associatedSites\": [\"https://iframe.com\"]}" https://iframe.com
```
that is to say `web.com` and `iframe.com` was in the same group.  And I also I disabled the third party cookies (screenshot)
<img width="321" alt="image" src="https://github.com/wanderview/quota-storage-partitioning/assets/32659870/9d4f7ea8-df09-4990-b4c0-659486ad9c73">

When I try to access the `iframe.com`'s cookie from `web.com` by using `requestStorageAccess`. I check the permission in this way
```
navigator.permissions.query({name: 'storage-access'}).then(res => {
      console.log(res.state);
      if (res.state === 'granted') {
        // Permission has already been granted
        // You can request storage access without any user gesture
        rSA();
      } else if (res.state === 'prompt') {
        // Requesting storage access requires user gesture
        // For example, clicking a button
        const btn = document.createElement("button");
        btn.textContent = "Grant access";
        btn.addEventListener('click', () => {
          // Request storage access
          rSA();
        });
        document.body.appendChild(btn);
      }
    });
```
The `res.state` is always the `prompt`.
<img width="460" alt="image" src="https://github.com/wanderview/quota-storage-partitioning/assets/32659870/3e2cffdb-0b8a-4687-8d60-bb8729416077">

Can you help me understand what happened? 
My expectation is that, we can access `iframe.com`'s cookies from `web.com` without using a `user gesture`. It's fine to call `requestStorageAccess` API to access the cookies, but I expected that I can always get the `state == grant` if I have implemented RSW.

Could you kindly suggest. Thank you in advanced!

**Comment by cfredric** [comment, marked as answer]
Hi @lqw1111, thanks for testing this!

> My expectation is that, we can access `iframe.com`'s cookies from `web.com` without using a `user gesture`. It's fine to call `requestStorageAccess` API to access the cookies, but I expected that I can always get the `state == grant` if I have implemented RSW.

This isn't quite how it works. RWS just allows Chrome to avoid having to show a storage access permission prompt in some cases; it doesn't preemptively grant storage access to the sites within the RWS. So in other words, I would expect the flow to look like:

1. Load `web.com` page with `iframe.com` iframe.
2. The iframe queries the `storage-access` permission.
    1. If the state is `granted`, the iframe immediately calls `document.requestStorageAccess()`.
    2. Otherwise the iframe waits for a user gesture and then calls `document.requestStorageAccess()`. (This is the same logic that you've implemented - it looks correct to me.)
3. Once the user clicks the button, Chrome will automatically grant the storage-access permission (without needing to prompt the user), and unblock cookie access within the iframe.

Can you confirm that that kind of flow works when you test it?

In the screenshot you provided, it looks like the iframe didn't wait for a user gesture before requesting storage access (given the `Must be handling a user gesture to use.` message). (Alternatively, maybe some of your other code had already consumed the user gesture by the time `document.requestStorageAccess()` was called; e.g. if you first opened a popup after the button click, then requested storage access.) So I think the `document.requestStorageAccess()` call was just missing the user gesture.

---

One other note - I'm guessing that the code snippet you gave is pseudocode (and isn't literally what your script executes), given the calls to `rSA()` instead of `document.requestStorageAccess()`. But just in case - you'll want to `await` the promise returned by `document.requestStorageAccess()`, otherwise the script might continue executing before the browser has finished granting storage access to the iframe.

**Reply by lqw1111** [reply]
Thank you @cfredric for the quick feedback.
Like you said, the flow works on my side, once the user grant the permission by using user gesture (in my case, clicking the button), then I am free to access the 3rd party cookies.

But my expectation is that, Is it possible that we can avoid request grant permission by user's gesture? Like the user always get the `grant` state rather than `prompt`. 
Because in my case, we can not stay here, say, Hi, please grant the permission to access 3rd party cookies, and let user click the button then we continue our rest flow.
Please kindly suggestion is there anyway to do that?

And one more question about the rollout of 3rd party cookies deprecation, once we rollout 100% users from Q3 2024. Does it means we can not use the 3rd party cookies anymore, even we have implement the RSW and using `requestStorageAccess` API and user gesture to grant the permission?

**Reply by cfredric** [reply]
> But my expectation is that, Is it possible that we can avoid request grant permission by user's gesture? Like the user always get the grant state rather than prompt.

Yes, that should work in Chrome Stable (118) today. The first time the user visits your page/iframe, the status will be `prompt` and you will need to request permission with a user gesture; but every visit after that (for at least 30 days), the status will be `granted` and you do not need a user gesture.

If you're seeing a different behavior during your testing, a screen-capture or demo would be helpful for me to investigate. But this works correctly for me using the [RWS demo page](https://fps-member-1.glitch.me/request-storage-access.html).

> And one more question about the rollout of 3rd party cookies deprecation, once we rollout 100% users from Q3 2024. Does it means we can not use the 3rd party cookies anymore, even we have implement the RSW and using requestStorageAccess API and user gesture to grant the permission?

Chrome will block third-party cookies by default, but it will still be possible to unblock them. I.e., sites will have to explicitly request (and be granted) access to them, e.g. via the Storage Access API, in order to use them. So, if you adopt the Storage Access API and Related Website Sets, you can expect those to continue to work in Chrome after Q3 2024.

**Reply by lqw1111** [reply]
> The first time the user visits your page/iframe, the status will be prompt and you will need to request permission with a user gesture; but every visit after that (for at least 30 days), the status will be granted and you do not need a user gesture.

Is it possible to avoid user gesture even at the first time visit? We'd like this grant permission steps is transparent to the user? Is it possible?

**Reply by cfredric** [reply]
> Is it possible to avoid user gesture even at the first time visit?

No, not in Chrome today. This is intentional, for cross-browser-interop. Other browsers don't support Related Website Sets, so in those browsers a user gesture is required and a permission prompt will be shown to the user. Since Chrome also requires a user gesture when permission hasn't been granted yet, your site is more likely to work in other browsers that support the Storage Access API, even without Related Website Sets.

**Reply by lqw1111** [reply]
I see. Thanks for explanation. @cfredric 
I found in the doc https://github.com/GoogleChrome/related-website-sets/blob/main/RWS-Submission_Guidelines.md#browser-behavior
> requestStorageAccess and requestStorageAccessFor will be auto-granted for up to five domains in the order listed within the associated subset category.

it mentioned that if we implement RSW, it will be auto-granted for up to five domains. I am curious does it means if we implement RSW, requestStorageAccess will be auto-granted. Does it means we can avoid user gesture even in the first time visit? Can you help understand what does it means.

Thank you!

**Reply by cfredric** [reply]
Good question - when we say "will be auto-granted", we mean that when your page requests storage access (after satisfying all the constraints Chrome requires of a new request), Chrome will grant the request automatically, without involving the user. It does not mean that Chrome preemptively requests storage access on your sites' behalf; your sites still need to explicitly request storage access, and still need to have the user gesture in order to do so (the first time).

**Reply by lqw1111** [reply]
Gotcha, that's super helpful, Thank you for the explanation!
